File: 백준/Silver/21736. 헌내기는 친구가 필요해/헌내기는 친구가 필요해.js
const fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().split('\n');

// 도연이가 만날 수 있는 사람의 수 출력 
// O: 빈공간, x:벽, I: 도연이 , P: 사람
function bfs(i,j){
    queue.push([i,j]);
    visited[i][j] = true
    let cnt = 0;  

    while(queue.length){
        let [x,y] = queue.shift();
        for(let k=0; k<4; k++){
            let nx = x+dx[k]
            let ny = y+dy[k]
            if(0<=nx && nx<n && 0<=ny && ny<m && !visited[nx][ny]){
                if(board[nx][ny] === 'O'){
                    queue.push([nx,ny]);
                    visited[nx][ny] = true
                }
                else if(board[nx][ny] === 'P'){
                    cnt += 1
                    queue.push([nx,ny]);
                    visited[nx][ny] = true
                } 
                else if(board[nx][ny] === 'X'){
                    continue;
                }
            }
        }
    }

    return cnt; 
}

const [n,m] = input.shift().split(' ').map(Number);
const board = input.map(v => v.split(''))

let visited = Array.from(Array(n),() => Array(m).fill(false));
let queue = [];

let dx = [-1,1,0,0];
let dy = [0,0,-1,1];
let answer = 0;

for(let i=0; i<n; i++){
    for(let j=0; j<m; j++){
        if(board[i][j] === 'I'){
            cnt = bfs(i,j);
            answer += cnt;
        }
    }
}

if(answer === 0){
    console.log('TT');
} else {
    console.log(answer);
}

File: 백준/Silver/21736. 헌내기는 친구가 필요해/헌내기는 친구가 필요해.py
import sys
input = sys.stdin.readline
from collections import deque

def bfs(i,j):
    queue.append([i,j])
    visited[i][j] = True
    cnt = 0

    while queue:
        x,y = queue.popleft()
        for k in range(4):
            nx = x+dx[k]
            ny = y+dy[k]
            if 0<=nx<n and 0<=ny<m and not visited[nx][ny]:
                if board[nx][ny] == 'O':
                    queue.append([nx,ny])
                    visited[nx][ny] = True
                elif board[nx][ny] == 'P':
                    cnt += 1
                    queue.append([nx,ny])
                    visited[nx][ny] = True
                elif board[nx][ny] == 'X':
                    continue
    
    return cnt

n,m = map(int,input().split())
answer = 0
board = [list(input().rstrip()) for _ in range(n)]

visited = [[False]*m for _ in range(n)]
queue = deque([])
dx = [-1,1,0,0]
dy = [0,0,-1,1]

for i in range(n):
    for j in range(m):
        if board[i][j] == 'I':
            cnt = bfs(i,j)
            answer += cnt

if answer == 0:
    print('TT')
else:
    print(answer)